### ISSUE TITLE
Xarray's explicit indexes

### ISSUE THREAD

**Issue by Huite** [Deltares/xugrid#35]
For examples, see:

https://github.com/martinfleis/xvec
https://github.com/dcherian/crsindex/blob/main/crsindex.ipynb

They can be set with:
https://docs.xarray.dev/en/stable/generated/xarray.DataArray.set_xindex.html

One issue is that in the UGRID conventions, the connectivity arrays are 2D. So the first dimension matches whatever's going on in the data (node, edge, face), but the second doesn't. This can be avoided with a trick using numpy structured arrays:

```python
import numpy as np
import xarray as xr

ds = xr.open_dataset("Grevelingen_FM_grid_20190603_0001_net.nc")
faces2d = ds["NetElemNode"].values.astype(int) - 1  # face node_connectivity
triangle = np.dtype([("0", int), ("1", int), ("2", int)])
faces = np.empty(len(faces2d), dtype=triangle)
faces.view(int)[:] = faces2d.ravel()
ds = ds.assign_coords(faces=("nNetElem", faces))
```

Using a structured array here is somewhat of a hack and it might not be futureproof?
An alternative approach is perhaps using a Pandas ExtensionArray. This seems like a lot of work: https://stackoverflow.com/questions/68893521/simple-example-of-pandas-extensionarray

Next, we can define a UgridIndex:

```python
class Ugrid2dIndex(Index):
    def __init__(
        self,
        node_x,
        node_y,
        faces,
    ):
        self._ugrid = xu.Ugrid2d(node_x.values, node_y.values, -1, faces.values.view(int).reshape((-1, 3)))

    @classmethod
    def from_variables(cls, variables, options):
        return cls(
            variables["NetNode_x"],
            variables["NetNode_y"],
            variables["faces"],
        )
```

This seems to do the trick:

```python
newds = ds.set_xindex(
    (
        "NetNode_x",
        "NetNode_y",
        "faces",
    ),
    Ugrid2dIndex,
)
```

Now we could start adding methods to the Ugrid2dIndex and make it possible to select values with `sel`. Or alternatively provide other operations via a `.ugrid` accessor.

However, it doesn't solve all problems that we encounter with UGRID topologies (?):

* Node, edge, and face dimension are fully separated according to xarray's data model. Selecting one but not the other will generally invalidate the data <-> topology link. (So solve via a `.ugrid.sel` and `.ugrid.isel` instead.)
* A full UGRID topology requires possibly connectivities with faces, edges, nodes, and possible interrelations. As far I can tell, the index is not carried around for every variable. (Solve by using multiple indexes internally containing full grid? One for faces, one for edges (with edge node connectivity). But seems impractical for data on nodes?)

I think the fundamental question is whether a UgridDataset is properly considered as an xarray Dataset: I don't think so, as it breaks a premise of fully separated, orthogonal dimensions. From that perspective, it may be better to accept this, and special case the UGRID topology dimensions.

**Comment by Huite** [comment]
Alternatively, maybe a Topology should be grouped together in a completely different way, see: https://xarray-datatree.readthedocs.io/en/latest/

**Comment by dcherian** [comment]
> Node, edge, and face dimension are fully separated according to xarray's data model. 

I think you can handle this in `Index.sel` which will be called by `Dataset.sel`

> One issue is that in the UGRID conventions, the connectivity arrays are 2D. So the first dimension matches whatever's going on in the data (node, edge, face), but the second doesn't. 

The same problem exists with handling CF style bounds arrays and `IntervalIndex`, so I think we have enough motivation to construct a proper solution in Xarray. I'm not entirely sure what that solution looks like (both structured and extension arrays have come up in the discussion).

**Comment by Huite** [comment]
> I think you can handle this in Index.sel which will be called by Dataset.sel

Yes, I think so too now. That would be a very comfortable solution.

> The same problem exists with handling CF style bounds arrays and IntervalIndex, so I think we have enough motivation to construct a proper solution in Xarray.

Ah, right -- that's good news!

Ideally, the same index could be associated with all three dimensions somehow (... I think). I think that's the only exceptional part of the UGRID topology versus the other uses of the explicit indexes.

I'm itching to find out what this entails, hopefully I can free up some time to investigate.

**Comment by Huite** [comment]
With the merging of https://github.com/pydata/xarray/pull/10137

```python

# %%

import xugrid as xu
import xarray as xr

from xarray import Index

# %%

class Ugrid2dIndex(Index):
    def __init__(
        self,
        node_x,
        node_y,
        faces,
    ):
        self._ugrid = xu.Ugrid2d(node_x.values, node_y.values, -1, faces.values.view(int).reshape((-1, 3)))

    @classmethod
    def from_variables(cls, variables, options):
        return cls(
            variables["mesh2d_node_x"],
            variables["mesh2d_node_y"],
            variables["mesh2d_face_nodes"],
        )

    def should_add_coord_to_array(self, name, var, dims):
        return True


# %%

ds = xu.data.elevation_nl(xarray=True)
ds2 = ds.set_coords("mesh2d_face_nodes")

newds = ds2.set_xindex(("mesh2d_node_x", "mesh2d_node_y", "mesh2d_face_nodes"), Ugrid2dIndex)
print(newds["elevation"])

```
<xarray.DataArray 'elevation' (mesh2d_nFaces: 5248)> Size: 21kB
[5248 values with dtype=float32]
Coordinates:
    mesh2d_face_x      (mesh2d_nFaces) float64 42kB ...
    mesh2d_face_y      (mesh2d_nFaces) float64 42kB ...
  * mesh2d_node_x      (mesh2d_nNodes) float64 22kB 9.926e+03 ... 3.217e+04
  * mesh2d_node_y      (mesh2d_nNodes) float64 22kB 3.847e+05 ... 3.974e+05
  * mesh2d_face_nodes  (mesh2d_nFaces, nmax_face) float64 126kB 138.0 ... 1.9...
Dimensions without coordinates: mesh2d_nFaces
Indexes:
  ┌ mesh2d_node_x      Ugrid2dIndex
  │ mesh2d_node_y
  └ mesh2d_face_nodes
Attributes:
    unit:     m NAP
```

This now succesfully propagates the indexes and the mesh2d_face_nodes coordinates to the DataArray, which is great news!

I noticted `.indexes` does still try to cast to PandasIndex, but I think this is enough to start prototyping!

**Comment by wpbonelli** [comment]
@Huite I think `.indexes` will stay that way since it is deprecated in favor of `.xindexes`, if I am reading this comment right? https://github.com/pydata/xarray/discussions/10257#discussioncomment-12956618

Excited to see where this goes.

**Comment by dcherian** [comment]
^ Yes exactly, use `.xindexes` instead.

> One issue is that in the UGRID conventions, the connectivity arrays are 2D. So the first dimension matches whatever's going on in the data (node, edge, face), but the second doesn't.

We can now defer to the index to decide which coord vars to include with a DataArray : https://github.com/pydata/xarray/pull/10137. This is released on `v2025.06.0`

**Comment by deltamarnix** [comment]
Since we are still on a 0.x version, I wouldn't mind to not deprecate, but directly change the API and make this the first 1.0 release. The ideas seem good and solve a bunch of issues.
This is the way we want to go forward forward and removing the old classes gives good errors to the users. Adding a bunch of warnings is often not so obvious, because people think warnings are not to be looked at.

**Comment by JoerivanEngelen** [comment]
> Since we are still on a 0.x version, I wouldn't mind to not deprecate, but directly change the API and make this the first 1.0 release. The ideas seem good and solve a bunch of issues. This is the way we want to go forward forward and removing the old classes gives good errors to the users. Adding a bunch of warnings is often not so obvious, because people think warnings are not to be looked at.

I see the removal of two very commonly used classes (UgridDataArray and UgridDataset) as quite a severe API change. I think if we want to push this to the 1.0 release and somewhat control damage for users, we need to think of the following as well:

- Have separate documentation for each xugrid release and a version switcher (e.g. [like pandas docs have](https://pandas.pydata.org/pandas-docs/stable/)). There will be packages depending on xugrid (e.g. iMOD Python) that will not update instantly to the 1.0 release, as the changing API of xugrid also affects the internals of these packages. Users of these packages are forced to use an older version of xugrid for a while, and thus need documentation of this older version as well. The 1.0 release docs will not be representative for them, given the API changes.
- Docs how to migrate xugrid 0.x code to 1.x code.

**Comment by dcherian** [comment]
Hello, we are looking to make a big splash at SciPy. I am currently working on building a gallery of examples for this stuff.

So I was curious about the state of your experiments with the indexes. 

Also, we are happy to help! Please reach out with questions if you have any.

**Comment by Huite** [comment]
Hey @dcherian!

Thanks for reaching out! The SciPy conference is in four days right?
I hadn't looked at for the last two weeks (and I really need to finish something else :grimacing: )...
But I just spent the last two hours and I've pushed a [draft PR](https://github.com/Deltares/xugrid/pull/373) starting to use the indexes seriously.

Basically, it seems to work incredibly well so far!

I've chucked the wrap module with the UgridDataArray and UgridDataset in the bin (which was by far the worst part of xugrid).
This following snippet runs with the changes in the PR:

```python
import xarray as xr
import xugrid as xu

# %%
# Grab some data with UGRID conventions, run .ugrid.initialize to create the UgridIndex inside the dataset:

ds = xu.data.elevation_nl(xarray=True)
uds = ds.ugrid.initialize()
uda = uds["elevation"]
uda.ugrid.plot()
```

I already had an "accessor" defined pending the changes to xarray's indexes; before it looked like this:

```python
class UgridDataArrayAccessor:
    def __init__(self, obj: xr.DataArray, grid):
        self.obj = obj
        self.grid = grid
```

Basically, I can replace this with a property that looks into the indexes:

```python
    def __init__(self, obj: xr.DataArray):
        self.obj = obj

    @property
    def grid(self) -> UgridType:
        indexes = list(self.obj.xindexes.values())
        grids = {index._ugrid for index in indexes if isinstance(index, UgridIndex)}
        if len(grids) != 1:
            raise ValueError("DataArray should contain exactly one UgridIndex")
        return grids.pop()
```

And many methods **just** work! (Like the plotting methods.)
For another test, I took a look at an interpolation function, I just needed to remove the UgridDataArray in the return statement (see the PR if you're interested).

There's some extra work in letting the initialize argument accept a `grid` argument and `grids` for Dataset; dispatching for different types of topology. Those are very straightforward changes.

It turns out the alternative constructors will also work beautifully; I hadn't realized static methods will also work on the accessor, so this'll work as well:

```python
uds = xr.Dataset.ugrid.from_geodataframe(gdf)
```

Which I find very pleasing!

There are probably a lot of small decisions to make relating to optional UGRID variables; obviously the `should_add_coord_to_array` shouldn't always return True, etc., but these all seem straightforward decisions.

So far, it seems like the explicit indexes are basically fully sufficient to support the goals of xugrid. Please let me know if I there's anything specific I can do to help with the gallery!

**Comment by benbovy** [comment]
I'm very excited to see this!

As Deepak says, don't hesitate to reach out if you have any question. There may be some subtleties on how the Index API is called in some common Xarray internal workflows... This is not documented yet.

> High: copy, equals, rename
   Low: sel, concat, join, reindex_like (probably in that order)

I think that you'll want `UgridIndex.isel()` as a high priority. Otherwise the whole index will be dropped as soon as you call `isel` or index one of the associated coordinate variables (this includes `sel` as well, which calls `isel` afterwards, unless you return new index objects and variables in the `IndexSelResult` object returned by `UgridIndex.sel()`) .

**Comment by benbovy** [comment]
> If I do a reduction along an additional dimension, it ends up dropping the Ugrid index, because at least one dimension isn't present in the variable that's returned by the reduction

Is it specific to DataArray? Within `_replace_maybe_drop_dims` we should also take `Index.should_add_coord_to_array()`  into account for setting `coords` before calling `filter_indexes_from_coords`... It looks like we forgot to support that case.

**Comment by Huite** [comment]
(From the linked xugrid issue, comment: https://github.com/Deltares/xugrid/issues/35#issuecomment-4014179437)

You asked:

> Is it specific to DataArray?

For a reduction, it does appear to be specific to DataArrays. If I comment out the face data variable:

```python
ds = xr.Dataset(
    {
        "node_data": xr.Variable(("nodes", "three"), np.random.rand(len(VERTICES), 3)),
#        "face_data": xr.Variable(("faces", "three"), np.random.rand(len(FACES), 3)),
    }
).assign_coords(coords)
```

Then `ds.mean("three")` does result in a Dataset with my custom Index.
However, if I try `ds[["node_data"]]` to create a new Dataset, it also ends up dropping the index.

> Within `_replace_maybe_drop_dims` we should also take `Index.should_add_coord_to_array()` into account for setting `coords` before calling `filter_indexes_from_coords`...

That sounds like the right fix! 👍🏼

**Comment by rsignell** [comment]
@Huite would it be useful to see what Claude Code comes up with for a plan to finish this off?
I'd be happy to give it a try, but I don't want to waste your time (which would be needed to review) either!  😟

**Comment by Huite** [comment]
Hi @rsignell,

I think @benbovy hit the nail on the head earlier:

> Is it specific to DataArray? Within `_replace_maybe_drop_dims` we should also take `Index.should_add_coord_to_array()` into account for setting `coords` before calling `filter_indexes_from_coords`... It looks like we forgot to support that case.

So it's really an xarray feature and would need xarray developer review. See also the [details in the xarray issue](https://github.com/pydata/xarray/issues/11215#issuecomment-4031817902).

With the write-up there, the problem is probably sufficiently described such that Claude can tackle it. I think my issue description there also provides a reasonable example for a test. At any rate, an xarray dev then needs to review (specifically, @benbovy, I reckon).
(I also could've submitted a PR like that myself, or let Claude do it, but I remain a bit pressed for time...)

Having said that, once that's landed in xarray, I can probably release an explicit indexes xugrid version in a day! -- the changes in xugrid are really quite succint.

**Comment by rsignell** [comment]
@Huite the xarray PR https://github.com/pydata/xarray/pull/11286 is what you needed, right?

**Comment by Huite** [comment]
@rsignell 

Yes, my progress is here: https://github.com/Deltares/xugrid/pull/373

No real issues so far, I just have to walk through the test suite and update them. I'm working on it off and on, I've just got two other deadlines at the moment.
